src/main/java/net/minecraftforge/common/ForgeRegistryAccessHooks.java

```diff
@@ -0,0 +1,222 @@
+     */
+    private static <T extends IForgeRegistryEntry<T>> void injectBuiltinElements(RegistryAccessExtension<T> extension, RegistryAccess.RegistryHolder builtin, RegistryResourceAccess.InMemoryStorage storage)
+    {
+        var registry = builtin.ownedRegistryOrThrow(extension.getRegistryKey());
+
+        for (var entry : registry.entrySet())
+        {
+            var value = entry.getValue();
+            int id = registry.getId(value);
+            var lifecycle = registry.lifecycle(value);
+            storage.add(builtin, entry.getKey(), extension.getCodec(), id, value, lifecycle);
+        }
+
+        printRegistryContents(INJECT, registry);
+    }
+
+    /**
+     * Creates a new RegistryAccess.RegistryHolder who's registries are populated with both vanilla's
+     * builtin registry elements and those registered to each RegistryAccessExtension. The returned
+     * RegistryHolder is used for cross-component reference lookups when inlining things such as
+     * Biome feature lists.
+     */
+    private static RegistryAccess.RegistryHolder createExtendedRegistryHolder(RegistryAccess.RegistryHolder vanillaBuiltin)
+    {
+        var extendedBuiltin = new RegistryAccess.RegistryHolder();
+        // Copy the vanilla builtin registries so that extension components can cross-reference vanilla components.
+        for (var data : RegistryAccess.knownRegistries())
+        {
+            copy(data.key(), vanillaBuiltin, extendedBuiltin);
+        }
+
+        // Copy the registered elements of each RegistryAccessExtension. These form the default/builtin values of the component.
+        for (var extension : ForgeRegistries.REGISTRY_ACCESS_EXTENSIONS)
+        {
+            copy(extension, extendedBuiltin);
+        }
+
+        return extendedBuiltin;
+    }
+
+    /*
+     * Helper for copying the entries of a Registry from one RegistryAccess to another.
+     */
+    private static <T> void copy(ResourceKey<? extends Registry<T>> key, RegistryAccess source, RegistryAccess dest)
+    {
+        var input = source.ownedRegistryOrThrow(key);
+        copy(key, input.entrySet(), dest);
+    }
+
+    /*
+     * Helper for copying the entries of a RegistryAccessExtension's Registry to a RegistryAccess.
+     */
+    private static <T extends IForgeRegistryEntry<T>> void copy(RegistryAccessExtension<T> extension, RegistryAccess dest)
+    {
+        var input = extension.getRegistry().get();
+        copy(extension.getRegistryKey(), input.getEntries(), dest);
+    }
+
+    /*
+     * Helper for copying an Iterable of ResourceKey-to-value mappings to a RegistryAccess.
+     */
+    private static <T> void copy(ResourceKey<? extends Registry<T>> key, Iterable<Map.Entry<ResourceKey<T>, T>> source, RegistryAccess dest)
+    {
+        var registry = dest.ownedRegistryOrThrow(key);
+        for (var entry : source)
+        {
+            registry.register(entry.getKey(), entry.getValue(), Lifecycle.stable());
```

Comment:
The lifecycle should be obtained from the elements itself. It should not just be declared as stable.

Reply:
Forge registries don't hold lifecycles for entries. Thoughts on stable or experimental for those? Will do for the vanilla ones